PYIC-7902/8614: allow multivalued journey context

thebauSoftwire opened the pull request:
## Proposed changes
### What changed

- allow multivalued journey context

### Why did it change

- we want a more flexible approach to journey contexts so that they can be used for future initiatives which may overlap

### Issue tracking
<!-- Jira ticket & other docs, like RFCs -->

- [PYIC-7902](https://govukverify.atlassian.net/browse/PYIC-7902)
- [PYIC-8614](https://govukverify.atlassian.net/browse/PYIC-8614)

## Checklists

- [ ] READMEs and documentation up-to-date
- [x] API/ unit/ contract tests have been written/ updated
- [x] No risk of exposure: PII, credentials, etc through logs/ code
- [ ] Production changes appropriately staged out


[PYIC-7902]: https://govukverify.atlassian.net/browse/PYIC-7902?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PYIC-8614]: https://govukverify.atlassian.net/browse/PYIC-8614?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

DanCorderIPV committed:
PYIC-7902: Switch journey context to a map

DanCorderIPV committed:
PYIC-7902: Add journey contexts to events and add ability to unset a context

thebauSoftwire committed:
Merge branch 'main' into PYIC-7902-experiment-extend-journey-context

# Conflicts:
#	api-tests/features/strategic-app/p2-strategic-app.feature
#	libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java

thebauSoftwire committed:
PYIC-8614: add unit tests for checkJourneyContext, journeyContextToSet and journeyContextToUnset

thebauSoftwire committed:
PYIC-8614: add unit tests for mergeSets and suppress sonarqube warning

thebauSoftwire commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/events/EventResolver.java:
```diff
@@ -54,11 +55,12 @@ private TransitionResult resolveExitNestedJourneyEvent(
         return resolve(exitEvent, parameters);
     }
 
+    @SuppressWarnings("java:S3776") // Refactor this method to reduce its Cognitive Complexity
```
Suppressing this warning as I don't think it's difficult to understand this method (it's got a high complexity score bc there's quite a few nested logic)

hugh-emerson commented on lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandlerTest.java:
```diff
@@ -1118,6 +1118,170 @@ void shouldReturn500IfCimitUtilityServiceThrowsExceptionWhenGettingMitigations(
+        spyIpvSessionItem.pushState(new JourneyState(INITIAL_JOURNEY_SELECTION, "PAGE_STATE"));
+        spyIpvSessionItem.setClientOAuthSessionId(SecureTokenHelper.getInstance().generate());
+        spyIpvSessionItem.setSecurityCheckCredential(SIGNED_CONTRA_INDICATOR_VC_1);
+        spyIpvSessionItem.setJourneyContext("someContext");
```
Why not use `IpvSessionItem spyIpvSessionItem = mockIpvSessionItemAndTimeout("PAGE_STATE")` instead of this big block?

You can still call `spyIpvSessionItem.setJourneyContext("someContext");` afterwards

thebauSoftwire replied:
ohh thanks! I thought I had to set it first before setting the mock

hugh-emerson reviewed (commented):
Discussed on Slack, but adding here for visibility: 

Needs to support both single and multivalued context at first for zero downtime. Single valued context can be removed in a subsequent release.

thebauSoftwire committed:
PYIC-8614: re-add support for IpvSessionItem.journeyContext

thebauSoftwire commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -752,14 +752,16 @@ states:
     events:
       uk:
         targetState: IDENTITY_START_PAGE
+        journeyContextToUnset: internationalAddress
       international:
         targetState: NON_UK_APP_INTRO_PAGE
+        journeyContextToSet: internationalAddress
         checkFeatureFlag:
           strategicAppEnabled:
             targetState: NON_UK_PASSPORT
+            journeyContextToSet: internationalAddress
 
   NON_UK_PASSPORT:
-    journeyContext: internationalAddress
     response:
       type: page
       pageId: non-uk-passport
```
I think this is fine to do, since whatever version of the process-journey-event lambda the user is on will have the associated journey map version so there shouldn't ever be a case where the user's state has a `journeyContext` but the lambda only supports `journeyContextToSet`

thebauSoftwire committed:
PYIC-8614: remove TODO comments

This was being flagged as a code smell by sonarqube. The cleanup just involves the IpvSessionItem file so cna remove the TODO reminders

thebauSoftwire committed:
PYIC-8614: reuse mockIpvSessionItemAndTimeout for unit test

hugh-emerson reviewed (dismissed):
Looks good to me 👍 

Did you test this is backwards compatible now? Eg:

* Deploy the current main to your dev environment
* Start a session and get a journey context (say internationalAddress)
* Deploy your changes
* Continue your journey with preferNoApp
* Check that you successfully land on the non-uk-no-app-options page

thebauSoftwire committed:
PYIC-8614: make journeyContexts a list instead of a set as DynamoDB does not support an empty set

hugh-emerson reviewed (approved):
👍
